### DISCUSSION TITLE
Leapp CLI & Core

### DISCUSSION THREAD

**Discussion by pethron** [Noovolari/leapp#52] [General]
As the codebase is expanding and we are trying to fit the needs of everyone while being open to the community and integrate with more and more third parties; we are starting to feel the need to do some separation of concerns of the application logic.

Namely, we would like to **split the core engine of Leapp into a separate daemon and a CLI on which the UI will be built upon** on a different project.

Like this, we think that we can better help both people using the UI and the terminal and empower people to easily get started in writing their own integrations and contribute to the project with a greater focus on their needs. 
It should come to even easier integration with package managers as brew and chocolatey, as pointed out in another discussion.

This thread would serve as both a discussion and poll to know if it's something you would like to see for Leapp.

**HOW TO VOTE**
To vote just react to this post with :+1: or :-1: to keep the discussion as clean as possible.

**Comment by andreacavagna01** [comment]
I think by splitting the daemon and the UX/UI we can also amplify the number of people that can contribute to the project.
Also, It could be a perfect time to make it more a plugin-based system, to help contributors bring their knowledge to the project.

**Comment by jgrumboe** [comment]
Is there any progress in this CLI discussion?
It would perfectly fit the use case to iterate over multiple accounts and do some aws-cli stuff in each account. Right now, we do have multiple profiles in `.aws/config` and use `export AWS_PROFILE=...` in a for-loop. But this seems not to work since with Leapp I have no profiles anymore. Or do I miss something?

**Reply by andreacavagna01** [reply]
The named-profile feature will be introduced in a future release soon, so if you are looking for them, wait a bit more. 👍 

Later on, the core logic of the Leapp will be inside the [Daemon](https://github.com/Noovolari/leapp-daemon) both the UI and the CLI project will work together locally with the Daemon.

What's your use case? Do you want to run the same script across multiple sessions programmatically? 
 
Also, your request seems to be linked to an enhancement of @sheldonhull
https://github.com/Noovolari/leapp-daemon/issues/20 
is that linked to that?

**Reply by jgrumboe** [reply]
Hmmm ... I've read through https://github.com/Noovolari/leapp-daemon/issues/20 but in fact I would be just satisfied without needing to deal with `.aws/config`.
For my use case it would be enough to have something like this (pseudo-code):
```
for account in $(leapp list-accounts --provider=aws);
do
  leapp active-account=$account
  aws s3 ls
done
leapp close-all-active-accounts
```
Something like that :)

**Reply by andreacavagna01** [reply]
That's a great idea, I think is a sort of "Action" that can run on one or multiple Session with a custom macro (in this case `aws s3 ls`).
I think AWS SSM Session is the starting point to generalize the idea of this action to make this kind of thing work!
